File: OTEAServer/Controllers/AddressesController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Http.HttpResults;
using Microsoft.AspNetCore.Mvc;
using OTEAServer.Misc;
using OTEAServer.Models;
using System.Data;
using System.Net;
using System.Text.Json;

namespace OTEAServer.Controllers
{
    ///<summary>
    /// Controller class for address operations
    /// Author: Pablo Ahíta del Barrio
    /// Version: 1
    /// </summary>


    [ApiController]
    [Route("Addresses")]
    public class AddressesController : ControllerBase
    {
        /// <summary>
        /// Database context
        /// </summary>
        private readonly DatabaseContext _context;


        /// <summary>
        /// Class constructor
        /// </summary>
        /// <param name="context">Database context</param>
        public AddressesController(DatabaseContext context)
        {
            _context = context;
        }

        /// <summary>
        ///  Method that obtains from the database all the addresses
        /// </summary>
        /// <returns>Addresses list</returns>
        [HttpGet("all")]
        [Authorize]
        public IActionResult GetAll([FromHeader] string Authorization)
        {
            try {
                var addresses = _context.Addresses.ToList();
                List<JsonDocument> result = new List<JsonDocument>();
                foreach (Address address in addresses)
                {
                    String rg = "{\"idAddress\":\"" + address.idAddress + "\"," +
                            "\"addressName\":\"" + address.addressName + "\"," +
                            "\"idCity\":\"" + address.idCity + "\"," +
                            "\"idProvince\":\"" + address.idProvince + "\"," +
                            "\"idRegion\":\"" + address.idRegion + "\"," +
                            "\"idCountry\":\"" + address.idCountry + "\"," +
                            "\"nameCity\":\"" + address.nameCity + "\"," +
                            "\"nameProvince\":\"" + address.nameProvince + "\"," +
                            "\"nameRegion\":\"" + address.nameRegion + "\"}";
                    result.Add(JsonDocument.Parse(rg));
                }
                return Ok(result);
            }
            catch (Exception ex) {
                return BadRequest(ex.Message);
            }
            
        }

        /// <summary>
        /// Method that obtains from the database the address using its identifier
        /// </summary>
        /// <param name="id">Address identifier</param>
        /// <returns>Address if is in the database, null if not</returns>
        [HttpGet("get")]
        [Authorize]
        public ActionResult<Address> Get([FromQuery] int id, [FromHeader] string Authorization)
        {
            try
            {
                var address = _context.Addresses.FirstOrDefault(a => a.idAddress == id);

                if (address == null)
                    return NotFound();

                return address;
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
            
        }

        /// <summary>
        /// Method that appends the address to the database
        /// </summary>
        /// <param name="address">New address</param>
        /// <returns>Address if was append, null if not</returns>
        [HttpPost]
        [Authorize]
        public IActionResult Create([FromBody] Address address, [FromHeader] string Authorization)
        {
            try
            {
                _context.Addresses.Add(address);
                _context.SaveChanges();
                return CreatedAtAction(nameof(Get), new { idAddress = address.idAddress }, address);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
            

        }


        /// <summary>
        /// Method that updates an address using its identifier
        /// </summary>
        /// <param name="idAddress">Address identifier</param>
        /// <param name="address">Address</param>
        /// <returns>Address if was updated, null if not</returns>
        [HttpPut]
        [Authorize]
        public IActionResult Update([FromQuery] int idAddress, [FromBody] Address address, [FromHeader] string Authorization)
        {
            try
            {
                if (idAddress != address.idAddress)
                    return BadRequest();

                var existingAddress = _context.Addresses.FirstOrDefault(a => a.idAddress == idAddress);
                if (existingAddress is null)
                    return NotFound();


                existingAddress.addressName = address.addressName;
                existingAddress.idCity = address.idCity;
                existingAddress.idProvince = address.idProvince;
                existingAddress.idRegion = address.idRegion;
                existingAddress.idCountry = address.idCountry;
                existingAddress.nameCity = address.nameCity;
                existingAddress.nameProvince = address.nameProvince;
                existingAddress.nameRegion = address.nameRegion;

                _context.SaveChanges();

                return Ok(existingAddress);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }

           
        }

        /// <summary>
        /// Method that deletes the address using its identifier
        /// </summary>
        /// <param name="id">Address identifier</param>
        /// <returns>Address if was deleted, null if not</returns>
        [HttpDelete]
        [Authorize]
        public IActionResult Delete([FromQuery] int id, [FromHeader] string Authorization)
        {
            try
            {
                var address = _context.Addresses.FirstOrDefault(a => a.idAddress == id);

                if (address is null)
                    return NotFound();

                _context.Addresses.Remove(address);
                _context.SaveChanges();

                return NoContent();
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
            
        }
    }
}

File: connection/src/main/java/otea/connection/controller/AddressesController.java
package otea.connection.controller;


import com.google.gson.JsonObject;

import java.io.IOException;
import java.net.SocketTimeoutException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import cli.organization.data.Address;
import otea.connection.ConnectionClient;
import otea.connection.api.AddressesApi;
import retrofit2.Call;
import retrofit2.Response;

/**
 * Controller class for address operations
 *
 * @author Pablo Ahíta del Barrio
 * @version 1
 *
 * */
public class AddressesController {

    /**Addresses API to connect to the server*/
    private static AddressesApi api;

    /**Controller instance*/
    private static AddressesController instance;

    private static int numAttempts=0;

    /**Class constructor*/
    private AddressesController(){
        api= ConnectionClient.getInstance().getRetrofit().create(AddressesApi.class);
    }

    /**
     * Method that obtains the controller instance
     *
     * @return Controller instance
     * */
    public static AddressesController getInstance(){
        if(instance==null){
            synchronized (AddressesController.class){
                if(instance==null){
                    instance=new AddressesController();
                }
            }
        }
        return instance;
    }

    /**Refresh API*/
    public static void refreshApi(){
        instance=new AddressesController();
    }

    /**
     * Method that obtains from the database all the addresses
     *
     * @return Addresses list
     * */
    public static List<Address> GetAll(){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<List<JsonObject>> callable = new Callable<List<JsonObject>>() {
            @Override
            public List<JsonObject> call() throws Exception {
                Call<List<JsonObject>> call = api.GetAll();
                Response<List<JsonObject>> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<List<JsonObject>> future = executor.submit(callable);
            List<JsonObject> list = future.get();
            executor.shutdown();
            numAttempts=0;
            List<Address> addresses=new ArrayList<>();
            for(JsonObject address:list){
                int idAddress=address.getAsJsonPrimitive("idAddress").getAsInt();
                String addressName=address.getAsJsonPrimitive("addressName").getAsString();
                int idCity=address.getAsJsonPrimitive("idCity").getAsInt();
                int idProvince=address.getAsJsonPrimitive("idProvince").getAsInt();
                int idRegion=address.getAsJsonPrimitive("idRegion").getAsInt();
                String idCountry=address.getAsJsonPrimitive("idCountry").getAsString();
                String nameCity=address.getAsJsonPrimitive("nameCity").getAsString();
                String nameProvince=address.getAsJsonPrimitive("nameProvince").getAsString();
                String nameRegion=address.getAsJsonPrimitive("nameRegion").getAsString();
                addresses.add(new Address(idAddress, addressName, idCity, idProvince, idRegion, idCountry, nameCity, nameProvince, nameRegion));
            }
            return addresses;
        } catch (InterruptedException | ExecutionException e) {
            if(e.getCause() instanceof SocketTimeoutException){
                numAttempts++;
                if(numAttempts<3) {
                    return GetAll();
                }
                else{
                    numAttempts=0;
                    return null;
                }
            }
            else{
                throw new RuntimeException(e);
            }
        }
    }

    /**
     * Method that obtains from the database the address using its identifier
     *
     * @param idAddress - Address identifier
     * @return Address if is in the database, null if not
     * */
    public static Address Get(int idAddress) {
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<Address> callable = new Callable<Address>() {
            @Override
            public Address call() throws Exception {
                Call<Address> call = api.Get(idAddress);
                Response<Address> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<Address> future = executor.submit(callable);
            Address result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that appends the address to the database
     *
     * @param address - New address
     * @return Address if was append, null if not
     * */
    public static Address Create(Address address) {
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<Address> callable = new Callable<Address>() {
            @Override
            public Address call() throws Exception {
                Call<Address> call = api.Create(address);
                Response<Address> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<Address> future = executor.submit(callable);
            Address result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that updates an address using its identifier
     *
     * @param id - Address identifier
     * @param address - Address
     * @return Address if was updated, null if not
     * */
    public static Address Update(int id, Address address){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<Address> callable = new Callable<Address>() {
            @Override
            public Address call() throws Exception {
                Call<Address> call = api.Update(id,address);
                Response<Address> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<Address> future = executor.submit(callable);
            Address result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that deletes the address using its identifier
     *
     * @param id - Address identifier
     * @return Address if was deleted, null if not
     *
     * */
    public static Address Delete(int id) {
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<Address> callable = new Callable<Address>() {
            @Override
            public Address call() throws Exception {
                Call<Address> call = api.Delete(id);
                Response<Address> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<Address> future = executor.submit(callable);
            Address result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }
}
